### PR TITLE
EdgeML Tensorflow [WIP]

### PR DESCRIPTION
**Pull request by adityakusupati**
The plug and play EdgeML classifiers in tensorflow.

### CODE REVIEW

File: tf/bonsai/bonsai.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+		self.totalNodes = 2*self.internalNodes + 1
+
+		## The Parameters of Bonsai
+		self.W = self.initW(W)
+		self.V = self.initV(V)
+		self.T = self.initT(T)
+		self.Z = self.initZ(Z)
+
+		## Placeholders for Hard Thresholding and Sparse Training
+		self.Wth = tf.placeholder(tf.float32, name='Wth')
```
Comment (metastableB) [inline tf/bonsai/bonsai.py]: @adityakusupati  We should move anything related to  training to a separate method and not make it part of the `__init__`. Sometimes, you'd just want to have a Bonsai subgraph initialized with pretrained weights for inference. For such cases, there is no reason to have these unused nodes (`Wth, Vth` etc) as part of the graph.

Also, `__init__` should also take the graph to attach to as an argument. What if I don't want Bonsai to be part of the global graph and instead am running multiple experimetns with different graphs ?

File: tf/bonsai/bonsai.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+	## Functions to initilaise Params (Warm start possible with given numpy matrices)
+	def initZ(self, Z):
+		if Z is None:
+			Z = tf.random_normal([self.projectionDimension, self.dataDimension])
+		Z = tf.Variable(Z, name='Z', dtype=tf.float32)
+		return Z
+
+	def initW(self, W):
+		if W is None:
+			W = tf.random_normal([self.numClasses*self.totalNodes, self.projectionDimension])
```
Comment (metastableB) [inline tf/bonsai/bonsai.py]: @adityakusupati I was wondering, should we allow a custom initializer as an optional argument instead of having a random normal initializer hardcoded? Like `LSTMCell` has?

Comment (adityakusupati) [reply tf/bonsai/bonsai.py]: The input param W, can be thought of as a custom initialiser. You can generate any intended initialisation and load it with W.

File: tf/bonsai/bonsai.py (hunk 1 above)
Comment (adityakusupati) [reply tf/bonsai/bonsai.py]: I agree with the first part. I need to think about it.

The second part is already addressed. feats is the argument which takes in the graph to be attached. I didn't get the doubt.

**Comment by metastableB** [comment]
@adityakusupati 
 I've restructure to the following

```
├── edgeml
│   ├── graph
│   │   ├── bonsai.py
│   │   ├── protonn.py # not included
│   ├── trainer
│   │   ├── bonsaiTrainer.py
│   │   ├── protoNNTrainer.py
│   └── utils.py # should probably be common/utils.py 
└── examples
    ├── bonsai_example.py
    ├── bonsaipreprocess.py
    ├── bonsaipreprocess.pyc
```

Please have a look. `bonsai_example.py` has now
```
from edgeml.trainer.bonsaiTrainer import BonsaiTrainer
from edgeml.graph.bonsai import Bonsai
```

File: tf/edgeml/trainer/bonsaiTrainer.py [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+        for i in range(totalEpochs):
+            print("\nEpoch Number: " + str(i))
+
+            trainAcc = 0.0
+            numIters = int(numIters)
+            for j in range(numIters):
+
+                if counter == 0:
+                    msg = " Dense Training Phase Started "
+                    print("\n%s%s%s\n" % (header, msg, header))
```
Comment (metastableB) [inline tf/edgeml/trainer/bonsaiTrainer.py]: Please accept a `files` argument to the print statement as part of your `trainer` routine. As things current stand, the only way for me to redirect console output is to redirect it using `>` .
`print` automatically accepts the `file` argument which can defaults to `None` and redirects to `sys.stdout`. Can you add an optional argument `outfilePointer` or something to your train routine or the class itself which defaults to `None`?

In case someone wants to redirect to file, he can provide a file pointer to do the same.

File: tf/edgeml/trainer/bonsaiTrainer.py [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+        self.Y = Y
+        self.X = X
+
+        self.useMCHLoss = useMCHLoss
+
+        self.learningRate = learningRate
+
+        self.assertInit()
+
+        self.score, self.X_ = self.bonsaiObj(self.X)
```
Comment (metastableB) [inline tf/edgeml/trainer/bonsaiTrainer.py]: @adityakusupati  I'm unsure how `tf`  will behave in the following case. Do you have an idea?

1. `bonsaiObj.__call__` was already called outside before passing to the trainer
2. `bonsaiTrainer` is called again on the same object.

In both cases, the first call to `__call__` will create the graph, but I'm unsure what happens when the call happens again (after the graph is already created).

File: tf/edgeml/trainer/bonsaiTrainer.py (hunk 3 above)
Comment (adityakusupati) [reply tf/edgeml/trainer/bonsaiTrainer.py]: I generally use the > and 2> redirection, but sure we can pass a file as an argument.

File: tf/edgeml/trainer/bonsaiTrainer.py (hunk 4 above)
Comment (adityakusupati) [reply tf/edgeml/trainer/bonsaiTrainer.py]: I think it re-writes the graph entirely as there are variables such as self.nodeProb etc., inherent to the object and they will be re-created. The old graph will not make any sense even though it is present in the current session as it has no binding even if it is present (ie., no path for backprop). I don't think we should worry about it as BonsaiTrainer is just an example routine (which is same as the one we used ICML). There are various other training routines like DSD which people can custom implement.

Try not to over think these. We are not tailor making EdgeML for every single purpose.

**Comment by adityakusupati** [comment]
@metastableB , I think the module structure is broken. It doesn't run out of the box. Is there a need to add various folders into the path before importing the modules for
from edgeml.trainer.bonsaiTrainer import BonsaiTrainer
from edgeml.graph.bonsai import Bonsai

**Comment by metastableB** [comment]
@adityakusupati 

1. How else would you do this? In terms of modules? The simplest would be `edgeml.bonsai.Bonsai` and `edgeml.bonsai.BonsaiTrainer` which would require you to add both `Bonsai` and `BonsaiTrainer` within the same file `bonsai.py`. I'm not in favor of that. 

2. Why do you think the current module structure is bad? It makes sense from my end to keep the trainer and graph separate.

File: tf/edgeml/trainer/bonsaiTrainer.py (hunk 4 above)
Comment (metastableB) [reply tf/edgeml/trainer/bonsaiTrainer.py]: I agree. I just was not sure how multiple graph calls interact with tf. If you think this is fine, i'm okay.

**Comment by adityakusupati** [comment]
All I meant to say was code isn't working out of the box, can you provide me with any special instructions?

@metastableB

**Comment by metastableB** [comment]
Add the following to your main script. 

```
import sys
sys.path.insert(0, '../')
```

I.e add the parent directory of `edgeml` to the PATH so that python can find the `edgeml` module.

**Comment by adityakusupati** [comment]
@metastableB , This works. I am was not sure about the setup.py thus the confusion.

File: tf/edgeml/trainer/bonsaiTrainer.py (hunk 3 above)
Comment (adityakusupati) [reply tf/edgeml/trainer/bonsaiTrainer.py]: I think this is fixed now. You can give an output file as an argument and get along with it. It defaults to sys.stdout.

Let me know if this is not fine

File: tf/edgeml/trainer/bonsaiTrainer.py [hunk 5]
```diff
@@ -37,6 +38,11 @@ def __init__(self, bonsaiObj, lW, lT, lV, lZ, sW, sT, sV, sZ,
 
         self.useMCHLoss = useMCHLoss
 
+        if outFile is not None:
```
Comment (metastableB) [inline tf/edgeml/trainer/bonsaiTrainer.py]: This is not needed actually. If file is `None` print will default to `sys.stdout` or whatever the default output stream is. Though there is no harm in doing this so its fine by me.

File: tf/examples/bonsai_example.py [hunk 6]
```diff
@@ -1,6 +1,9 @@
 import bonsaipreprocess
 import tensorflow as tf
 import numpy as np
+import sys
+sys.path.insert(0, '../')
```
Comment (metastableB) [inline tf/examples/bonsai_example.py]: This should not be part of bonsai example. Can you please remove this? We will add this to `README.md` or something

File: tf/examples/bonsaipreprocess.py [hunk 7]
```diff
@@ -80,6 +80,8 @@ def getArgs():
                         help='Sparsity for branching parameter Theta  (default: For Binary classification 1.0 else 0.2 try: [0.1, 0.3, 0.5])')
     parser.add_argument('-sZ', type=checkFloatPos, default=0.2,
                         help='Sparsity for projection parameter Z  (default: 0.2 try: [0.1, 0.3, 0.5])')
+    parser.add_argument('-oF', '--output_file', default=None,
```
Comment (metastableB) [inline tf/examples/bonsaipreprocess.py]: Can you change the helper text to 'Output file for dumping the program output, defaults to `stdout`'

**Comment by metastableB** [comment]
@adityakusupati  I think Bonsai looks solid now? You just have some minor changes to make and include a readme for downloading CURET and USPS data?

Can you review `protoNN.py` and `protoNNTrainer.py` ?

File: tf/edgeml/trainer/bonsaiTrainer.py (hunk 5 above)
Comment (adityakusupati) [reply tf/edgeml/trainer/bonsaiTrainer.py]: @metastableB Yes, but .flush() method works on sys.stdout and not on None.
I need to ensure that it happens.
We need to flush every epoch stats immediately to the files, which is not done by default, thus had to do it this way.

File: tf/examples/bonsai_example.py (hunk 6 above)
Comment (adityakusupati) [reply tf/examples/bonsai_example.py]: I know, I was testing it and forgot to remove it. Let it stay for our tests until you get steup.py up and running.

File: tf/examples/bonsaipreprocess.py (hunk 7 above)
Comment (adityakusupati) [reply tf/examples/bonsaipreprocess.py]: Sure.

**Comment by adityakusupati** [comment]
@metastableB , I haven't removed the sys.path in bonsai_example as of now, will do it at the end when we have the package ready. I shall start reviewing the protoNN codebase.
@AIgen , do have a look at protoNN if you wish to

File: tf/edgeml/graph/protoNN.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+        projectionDimension: (model paramter)
+        numPrototypes: (model parameter)
+        numOutputLabels: The number of output labels or classes
+        W, B, Z: initial value to use for W, B, Z
+        Expected Dimensions:
+        W   d x d_cap
+        B   d_cap x m
+        Z   L x m
+        X   n x d
+        '''
```
Comment (adityakusupati) [inline tf/edgeml/graph/protoNN.py]: Change the d, d_cap to the the elongated variable names like inputDimension etc.,

File: tf/edgeml/graph/protoNN.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        with tf.name_scope('protoNN') as ns:
+            self.__nscope = ns
+        self.__d = inputDimension
+        self.__d_cap = projectionDimension
+        self.__m = numPrototypes
+        self.__L = numOutputLabels
+
```
Comment (adityakusupati) [inline tf/edgeml/graph/protoNN.py]: Change the variable names to more readable ones like inputDims, projDims etc.,
try not to use single letters as above

Comment (metastableB) [reply tf/edgeml/graph/protoNN.py]: Variables with `__` are internal and private. I'll change it anyway.

File: tf/edgeml/trainer/protoNNTrainer.py [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+            Y [-1, num Labels]
+        '''
+        self.protoNNObj = protoNNObj
+        self.__regW = regW
+        self.__regB = regB
+        self.__regZ = regZ
+        self.__sW = sparcityW
+        self.__sB = sparcityB
+        self.__sZ = sparcityZ
+        self.__lR = learningRate
```
Comment (adityakusupati) [inline tf/edgeml/trainer/protoNNTrainer.py]: All the reg, spar and learning rate should be public, THis will help the users change them before re-running the same trainer again (retrain from the old solution)

File: tf/edgeml/trainer/protoNNTrainer.py [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+        self.__validInit = True
+        return True
+
+    def __lossGraph(self):
+        pnnOut = self.__protoNNOut
+        l1, l2, l3 = self.__regW, self.__regB, self.__regZ
+        W, B, Z, _ = self.protoNNObj.getModelMatrices()
+        with tf.name_scope('protonn-l2-loss'):
+            loss_0 = tf.nn.l2_loss(self.Y - pnnOut)
+            reg = l1 * tf.nn.l2_loss(W) + l2 * tf.nn.l2_loss(B)
```
Comment (adityakusupati) [inline tf/edgeml/trainer/protoNNTrainer.py]: Call it l2_loss

File: tf/edgeml/trainer/protoNNTrainer.py [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+            reg += l3 * tf.nn.l2_loss(Z)
+            loss = loss_0 + reg
+        return loss
+
+    def __trainGraph(self):
+        with tf.name_scope('protonn-gradient-adam'):
+            trainStep = tf.train.AdamOptimizer(self.__lR)
+            trainStep = trainStep.minimize(self.loss)
```
Comment (adityakusupati) [inline tf/edgeml/trainer/protoNNTrainer.py]: Don't re-write the trainStep variable, use it only once.

File: tf/edgeml/trainer/protoNNTrainer.py [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+        assert totalEpochs >= 1, 'Total epochs should be psotive integer'
+        assert x_train.ndim == 2, 'Expected training data to be of rank 2'
+        assert x_train.shape[1] == d, 'Expected x_train to be [-1, %d]' % d
+        assert x_test.ndim == 2, 'Expected testing data to be of rank 2'
+        assert x_test.shape[1] == d, 'Expected x_test to be [-1, %d]' % d
+        assert y_train.ndim == 2, 'Expected training labels to be of rank 2'
+        assert y_train.shape[1] == L, 'Expected y_train to be [-1, %d]' % L
+        assert y_test.ndim == 2, 'Expected testing labels to be of rank 2'
+        assert y_test.shape[1] == L, 'Expected y_test to be [-1, %d]' % L
+
```
Comment (adityakusupati) [inline tf/edgeml/trainer/protoNNTrainer.py]: Move these to a new function, assertTrain()

File: tf/edgeml/trainer/protoNNTrainer.py [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+                        X: batch_x,
+                        Y: batch_y
+                    }
+                    acc_, loss_ = sess.run([self.accuracy, self.loss],
+                                           feed_dict=feed_dict)
+                    acc += acc_
+                    loss += loss_
+                acc /= len(y_test_batches)
+                loss /= len(y_test_batches)
+                print("Test Loss: %2.5f Accuracy: %2.5f" % (loss, acc))
```
Comment (adityakusupati) [inline tf/edgeml/trainer/protoNNTrainer.py]: file=redirFile missing

File: tf/examples/protoNN_example.py [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+    LEARNING_RATE = 0.1
+    # -----------------
+    # End configuration
+    # -----------------
+    out = loadData(DATA_DIR)
+    dataDimension = out[0]
+    numClasses = out[1]
+    x_train, y_train = out[2], out[3]
+    x_test, y_test = out[4], out[5]
+
```
Comment (adityakusupati) [inline tf/examples/protoNN_example.py]: Write an arg parser

**Comment by adityakusupati** [comment]
@metastableB , I was not able to find kmeans based initialisation as part of ProtoNN trainer. Is it by choice?

File: tf/edgeml/trainer/protoNNTrainer.py (hunk 10 above)
Comment (metastableB) [reply tf/edgeml/trainer/protoNNTrainer.py]: Good point. I'll have `getSparcity`, `setSparcity` methods. Directly writing to variables is a no no.

**Comment by adityakusupati** [comment]
@metastableB , are almost done? Did you test IHT routine?

**Comment by metastableB** [comment]
@adityakusupati Nope. Will do post May 18.

**Comment by metastableB** [comment]
@adityakusupati  Rebasing with master, master has diverged far too much.
